src/node_file.cc

```diff
@@ -1150,28 +1153,133 @@ static void RMDir(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+int MKDirpSync(uv_loop_t* loop, uv_fs_t* req, const std::string& path, int mode,
+               uv_fs_cb cb = nullptr) {
+  FSContinuationData continuation_data = FSContinuationData(req, mode, cb);
+  continuation_data.PushPath(path);
+
+  while (continuation_data.paths.size() > 0) {
+    std::string next_path = continuation_data.PopPath();
+    ssize_t err = uv_fs_mkdir(loop, req, next_path.c_str(), mode, nullptr);
+    switch (err) {
+      case 0:
+        break;
+      case UV_ENOENT: {
+        std::string dirname = next_path.substr(0, path.find_last_of("\\/"));
+        continuation_data.PushPath(next_path);
+        continuation_data.PushPath(dirname);
+        break;
+      }
+      case UV_EPERM: {
+        return err;
+        break;
+      }default:
+        err = uv_fs_lstat(loop, req, next_path.c_str(), nullptr);
+        if (err < 0) return err;
+        if (!S_ISDIR(req->statbuf.st_mode)) return UV_EEXIST;
+        break;
+    }
+  }
+
+  return 0;
+}
+
+int MKDirpAsync(uv_loop_t* loop,
+                uv_fs_t* req,
+                const char* path,
+                int mode,
+                uv_fs_cb cb) {
+  FSReqBase* req_wrap = static_cast<FSReqBase*>(req->data);
+  // on the first iteration of algorithm, stash state information.
+  if (req_wrap->continuation_data == nullptr) {
+    req_wrap->continuation_data = std::unique_ptr<FSContinuationData>{
+                                         new FSContinuationData(req, mode, cb)};
+    req_wrap->continuation_data->PushPath(path);
+  }
+
+  // on each iteration of algorithm, mkdir directory on top of stack.
+  std::string next_path = req_wrap->continuation_data->PopPath();
+  ssize_t err = uv_fs_mkdir(loop, req, next_path.c_str(), mode,
+                        uv_fs_callback_t{[](uv_fs_t* req) {
```

Comment:
Can you use `req_wrap->Dispatch(uv_fs_mkdir, next_path.c_str(), mode, uv_fs_callback_t{ …` here? We need that to keep track of how many requests are active/inactive so that we know when it’s safe to shut down the event loop

(ditto for `uv_fs_lstat` below)

Reply:
I *think* we need to call `uv_fs_req_cleanup()` after a `uv_fs_mkdir` call is done, but before we start the next one?

Reply:
I'm now calling `uv_fs_req_cleanup()` for all but the last `fs` operation, since it's called in the codebase already at the top level of the request wrap.